packages/webamp/js/components/Vis.tsx

```diff
@@ -0,0 +1,195 @@
+  }, [analyser, analyser.fftSize]);
+  const colors = useTypedSelector(Selectors.getSkinColors);
+  const mode = useTypedSelector(Selectors.getVisualizerStyle);
+  const audioStatus = useTypedSelector(Selectors.getMediaStatus);
+  const getWindowShade = useTypedSelector(Selectors.getWindowShade);
+  const getWindowOpen = useTypedSelector(Selectors.getWindowOpen);
+  isMWOpen = getWindowOpen("main");
+  doubled = useTypedSelector(Selectors.getDoubled);
+  const dummyVizData = useTypedSelector(Selectors.getDummyVizData);
+  const toggleVisualizerStyle = useActionCreator(Actions.toggleVisualizerStyle);
+  const windowShade = getWindowShade("main");
+
+  smallVis = windowShade && isMWOpen;
+  const renderWidth = 76;
+  const renderWidthBG = !isMWOpen 
+  ? renderWidth 
+  : windowShade 
+    ? (doubled ? renderWidth : 38)
+    : renderWidth * PIXEL_DENSITY;
+
+  renderHeight = smallVis ? 5 : 16;
+  PIXEL_DENSITY = doubled && smallVis ? 2 : 1;
+
+  const width = renderWidth * PIXEL_DENSITY;
+  const height = renderHeight * PIXEL_DENSITY;
+
+  const bgCanvas = useMemo(() => {
+    return preRenderBg(
+      renderWidthBG,
+      height,
+      colors[0],
+      colors[1],
+      Boolean(windowShade)
+    );
+  }, [colors, height, width, windowShade]);
+
+  const [canvas, setCanvas] = useState<HTMLCanvasElement | null>(null);
+
+  //? painter administration
+  const [painter, setPainter] = useState<VisPaintHandler | null>(null);
+
+  useEffect(() => {
+    if (!canvas) return;
+    const _setPainter = (PainterType: typeof VisPaintHandler) => {
+      const _vis: IVis = {
+        canvas,
+        colors,
+        analyser,
+        oscStyle: "lines",
+        bandwidth: "wide",
+        coloring: "normal",
+        peaks: true,
+        saFalloff: "moderate",
+        saPeakFalloff: "slow",
+        sa: "analyzer",
+      };
+      const newPainter = new PainterType(_vis);
+      setPainter(newPainter);
+    };
+    switch (mode) {
+      case VISUALIZERS.OSCILLOSCOPE:
+        _setPainter(WavePaintHandler);
+        break;
+      case VISUALIZERS.BAR:
+        _setPainter(BarPaintHandler);
+        break;
+      case VISUALIZERS.NONE:
+        _setPainter(NoVisualizerHandler);
+        break;
+      default:
+        _setPainter(NoVisualizerHandler);
+    }
+  }, [analyser, canvas, mode, colors]);
+
+  useEffect(() => {
+    if (canvas == null || painter == null) {
+      return;
+    }
+
+    const canvasCtx = canvas.getContext("2d");
+    if (canvasCtx == null) {
+      return;
+    }
+    canvasCtx.imageSmoothingEnabled = false;
+
+    let animationRequest: number | null = null;
+
+    const loop = () => {
+      if (mode === VISUALIZERS.NONE) {
+        canvasCtx.clearRect(0, 0, renderWidthBG, height);
+      } else {
+        canvasCtx.drawImage(bgCanvas, 0, 0);
+      }
+      painter.prepare();
+      painter.paintFrame();
+      animationRequest = window.requestAnimationFrame(loop);
+    };
+
+    if (audioStatus === MEDIA_STATUS.PLAYING) {
```

Comment:
Maybe we could bail out much earlier if `audioStatus !== MEDIA_STATUS.PLAYING`? I don't see any reason to create the canvas context if we are not playing since we don't actually render anything.

Reply:
This one I can't figure out without breaking everything

Reply:
Okay, I can come back and try to refactor this after it lands.